### ISSUE TITLE
Move from LDAP to userdb

### ISSUE THREAD

**Issue by RyanSquared** [hashbang/hashbang#1]
@KellerFuchs and @lrvick both want to move away from LDAP (openldap to be specific). This issue serves as a general purpose discussion forum to discuss the possible resolutions.

**Current Roadmap:** Fully automate and document the following:

- [ ] Migrate data from LDAP to a userdb instance
- [ ] Set up `userdb` read-only copy on shell servers
- [ ] Install and set configuration for nss-pgsql
- [ ] Create new `ssh_authorized_keys_command`
- [ ] Rewrite hashbangctl to interact with postgresql
  - **Note:** We're completely rewriting it from the ground up

**Comment by lrvick** [comment]
Potential solution: Replace provisor with a daemon that meets the following criteria:
- Would run on every shell server
- Provides socket interface to local machine to read/write to user data. (For consumption by hashbangctl and potentially other things)
- Provides RFC1288 (finger) interface for fetching user profiles via .plan,.project,.pgpkey files
- Provides REST interface to outside world that exposes server stats, read-only user-data, and user-creation endpoints
- Remote user creation service could hit Provisor on any shell server (perhaps round-robin DNS alias this at api.hashbang.sh)
- Admins can write any data, users can write to their own data. anyone can read
- Backed by a predictable, maintainable database such as PostgreSQL
- Must provide some interface postfix can consume for mail routing
- Must provide some interface sshd can consume for user -> public key resolution
- Must somehow provide expected system user interfaces like sssd does. (This is the part that is most unclear to me)
- `make test` should be a thing from day one so this stays a solid, easy-to-understand/maintain/extend part of our infra. (And hopefully used by others outside #!)
- has a descriptive name/brand that describes its purpose in maintaining a network of shell users. I submit "shelld" where hashbangctl becomes "shellctl"

**Comment by lrvick** [comment]
@KellerFuchs @ChickenNuggers @daurnimator  ^ would love your input here (or anyone for that matter)

**Comment by RyanSquared** [comment]
> Provides REST interface to outside world that exposes server stats, read-only user-data, and user-creation endpoints

Sounds great. But if you make it REST I will burn you alive. You can just as easily build a queryable JSON-formatted API without trying to build a REST-compatible clusterfuck.

> has a descriptive name/brand that describes its purpose in maintaining a network of shell users. I submit "shelld" where hashbangctl becomes "shellctl"

How original. I was thinking we could keep with `hashbangctl` and go with `hashbangd` or something at least with a bit of creativity behind it.

Also,

> Provides socket interface to local machine to read/write to user data. (For consumption by hashbangctl and potentially other things)

contradicts

> Remote user creation service could hit Provisor on any shell server (perhaps round-robin DNS alias this at api.hashbang.sh)

**Comment by lrvick** [comment]
@ChickenNuggers Simple JSON api of some kind.

Also no contradiction. A daemon can be accessible by socket/http at the same time.

The reason I was not suggesting hashbangctl/hashbangd is they are very -hashbang- specific. I think many other people might want to use something like this at universities, hacker-spaces, etc. It is easier to convey this is portable if our name is not burned into it, imo. I am thinking it be a hashbang-sponsored project critical to our infrastructure, that anyone can grab and run in theirs.

**Comment by RyanSquared** [comment]
If we're making a "simple" JSON API, using Flask's builtin `GET`, `PUT`, `PATCH`, etc. would be a better option than to use an API.

**Comment by KellerFuchs** [comment]
Oops, my comment somehow got lost.

@lrvick I think the following points are scope-creep, and have nothing to do with the auth system:
1. _Provides RFC1288 (finger) interface for fetching user profiles via .plan,.project,.pgpkey files_
2. _Provides REST interface to outside world that exposes server stats, read-only user-data_

I agree we want that, but it really shouldn't be the same code.
I have no problem with the deamon implementing (2) being coupled with the auth system (as in knowing the database schema, for instance).

You also set a requirement I disagree with:
- _Remote user creation service could hit Provisor on any shell server_  
  Do we really want to require having a multi-master DB?

Also, like @ChickenNuggers I'm pretty fond of the name `hashbangctl`/`hashbangd` (or `hashbangctld` ?), but let's not start bikesheding now.,

All in all, I think we want to separate a bunch of components here:
- The backing store, probably a distributed DB of some sort.  
  I think it should support having “read only” instances: things like the mail server should not be able to write there.
- `hashbangctld`, the daemon that mediates the access to the backing store.  
  It listens on a local UNIX socket, uses `SO_PEERCRED` to authenticate the user connecting to it, and speaks [a simple JSON-based protocol](https://gist.github.com/KellerFuchs/3ccdebe1fa466f55ffb5) over this.
- Integration with individual services (SSH, Postfix, IRC (?), ...) for authentication  
  For instance, the SSH integration can be a simple script (configured to be the `AuthorizedKeysCommand`) that connects to `hashbangctld`, retreives the data associated with a given user, and extracts the keys from it.
- Integration with individual services for “configuration”  
  Some services, like DNS or Postfix, need to pull information from the user db.
- A PAM module for authenticating the admins (when using `sudo(8)`)  
  This could be independent from `hashbangctld` and use static data (in `/etc` or managed through Ansible), in principle.
- `nss(5)` integration, for providing the `passwd` database.

**Comment by lrvick** [comment]
@KellerFuchs I am not sure how you are getting I was suggesting a mult-master DB. I didn't state it but in my mind the sane way to deploy the database would be external to any of the shell-boxes imo. N daemons talk to one central database. I would propose a standard master/slave configuration across regions where the slave can be promoted to master at any time the master becomes unavailable. If we want full distributed we may want to look into something like Consul instead.

Also totally fair the finger service probably should be its own standalone thing, but could simply talk to one of the daemons via http or socket. This type of separation of concerns allows for efforts like these to happen in parallel by different people. This is out of scope and can easily be written on top of this groundwork if we do it right. Instead I think it would be more helpful to define interfaces our user daemon offers based on the types of clients we forsee consuming it (hashbangctl, sshd, fingerd, postfix, etc). 

I also feel if we do this correctly, clients should have no knowledge at all of what the backing database is or how to use it, or credentials to reach it. We could instead expose a socket anyone can read, but users can only write to via hashbangctl on behalf of themselves. Not clear on what implementation details make the most sense here but there are certainly options.

Can you explain why we would want Postfix/DNS to talk directly to the DB instead of being abstracted via socket/http interfaces? Imo the clients should not need to know our database type, schema or anything, and we should be able to gut the database and swap in another one down the road, without breaking anything else. Or implement clever caching, etc. I feel we give up all control once clients know about the DB directly.

re bikeshedding: shellctl is just one suggestion I don't have strong feelings on but I do feel like hashbangd/hashbangctl will get us less external adoption from a branding perspective, but we can be democratic about this and If no one feels the same as me then I will let it be. If put to a vote with the active admins, daurnimator gets the deciding vote... or ties us.

I am very interested in details on how we might go about implementing the pam/nss integrations.

Super stoked about the direction this discussion is going. Hopefully we can get a draft design document started out of this...

**Comment by KellerFuchs** [comment]
@lrvick
- _“I am not sure how you are getting I was suggesting a mult-master DB”_
  _“Remote user creation service could hit Provisor on any shell server”_ sounds like writes can be done anywhere ...
- _“Can you explain why we would want Postfix/DNS to talk directly to the DB”_
  I don't ...
  I said _“I have no problem with the deamon [providing a REST interface to outside world that exposes server stats & read-only user-data] being coupled with the auth system (as in knowing the database schema, for instance).”_.   That's not about either Postfix or DNS. (And we would want that because otherwise, `hashbangctld` has to do the stats aggregation, and that's out of scope for it)

**Comment by lrvick** [comment]
You write to any daemon and any daemon knows the shared database location and can write changes to it.

**Comment by KellerFuchs** [comment]
@lrvick Why is that something we want?
(Hint: that's how you build things that can commit split-brain or have inconsistent histories ...)

**Comment by lrvick** [comment]
Basically all operations that touch users go through this gateway daemon which can be backed by any type of database we want hosted anywhere. Only this daemon would have write access to it

**Comment by lrvick** [comment]
How would you split brain?  This is how most horizontally scalable apis work. N API nodes act as the gsteway for one seperate database cluster.

**Comment by KellerFuchs** [comment]
@lrvick See `#!social`.  Basically, there was a misunderstanding between you saying “hashbangctld should run on all servers and route write queries” and me understanding “hashbangctld should run on all servers with a copy of the DB and handle write queries”.

I would still argue, there, that having hashbangctld expose an API, over the network, for writing stuff to the DB (or reading from it) is not what we want: our own services can run their own local hashbangctld.

Also, having a read-only hot replica of the DB on each server makes sense: we can safely handle a transient failure of the master, and it speeds up reads tremendously.

**Comment by lrvick** [comment]
Sure, we would have the ability to run local hashbangd's as we like. The idea here is lending maximum flexibility while also making sure we make it easy for people to build things on our platform. Our brittle insecure user creation server we probably don't want to trust with database credentials and it can just use a dumb http create like it does today.

More demanding use cases where we trust the server, can/should use the socket. But for someone making a simple stats widget or something, irc bot... trival use cases.. an http api gets the job done.

**Comment by KellerFuchs** [comment]
@lrvick Then it should be something separate from `hashbangd`, IMHO.

Regarding stats, I still argue that's not something you want to do over hashbangd's API: hashbangd is about doing queries about a specific user, whereas stats require aggregate queries.
That kind of mismatch is why the current stats thing is slow as hell, BTW: it iterates over all users.

**Comment by lrvick** [comment]
Well I was assuming the scope of hashbangd would provide all our general endpoints for all our common user CRUD operations via socket or http. In doing so no other services would ever need to know the database exists and it minimizes technical debt from what we may later find is a poor database choice.

**Comment by lrvick** [comment]
Also, for security. Only one API source means only one codebase that gets to see database credentials putting less security burden on downstream clients. Downstream clients then can be given only the very specific operations we choose to expose, instead of "here is the entire database, do whatever". Further, this lets us implement caching for things like the stats endpoints, and put less direct stress on the database. "I just served a stats request a second ago... here just take that one"

**Comment by KellerFuchs** [comment]
> hashbangd would provide all our general endpoints for all our common user CRUD operations

Exactly.  Stats isn't CRUD ...
Also, it's definitely possible to give specific services credentials that only allow specific operations (hell, pgsql even has row-level ACLs :P)

**Comment by lrvick** [comment]
This is true. I guess it just depends on  if we want to commit to pgsql forever. It is common wisdom to hide your database apis else you become married to them... pros and cons I suppose.

**Comment by KellerFuchs** [comment]
@lrvick I don't see how splitting “CRUD access to users” and “HTTP API exposing some stats” in two daemons doesn't hide the DB.

**Comment by lrvick** [comment]
Also client-side web-applications and simple bash scripts users create should be able to interact with our services without knowing secrets. I suppose we could publish a read-only set of credentials, or even allow a global limited read-only access to anyone. But then everyone knows our schema, and we can never change it.

**Comment by KellerFuchs** [comment]
@lrvick I never, ever, ever suggested that all applications should use the DB.  Please read what I write instead of putting words in my mouth.  That's one of the most infuriating things you tend to do  :(

**Comment by lrvick** [comment]
You indicated > 1 codebase would know about the DB which is where my hesitation is founded. Just trying to feel out where you are coming from and what the overhead gains us.

**Comment by lrvick** [comment]
In any event from what I am seeing we are mostly in agreement that a socket API and a HTTP api could cover our use cases via over-the-network or local depending on the use case.

The question are around pros/cons of building them as a single codebase/daemon or two? I vote one as then we have limited code surface to consider when making database changes. Happy to hear more counter arguments.

**Comment by lrvick** [comment]
If 2 daemons, then we probably want a library both share to abstract db operations, like provisor does for ldap today. This might be the maintainbale way forward if we go with this approach.

I just mainly want to avoid >1 codebase knowing the db schema. There is no way we get everything right with that the first time....

**Comment by lrvick** [comment]
This is what gave me the impression you were considering the possibility of N services with direct DB access: "Also, it's definitely possible to give specific services credentials that only allow specific operations (hell, pgsql even has row-level ACLs :P)"

I now see you were only really considering the individual needs of only the potentially 2 daemons here. My bad.

**Comment by KellerFuchs** [comment]
@lrvick I don't get why you seem to believe that it's easier to maintain the same amount of code (that knows about the schema) if it is “a single codebase”, knowing that it is under our control.

On the other hand, the cons are:
- all instances must be updated, even those that are not concerned by the change in the code;
- it's harder to know which codepaths can be exerciced by a give instance (making debug harder);
- we either end up with instances running with full access to the DB when they don't need to, or with methods returning “this instance doesn't have enough permissions” (or pretending not to have that method, which is pretty much the same).

**Comment by lrvick** [comment]
My first impression was each daemon repo would directly contain the code responsible for DB interaction, but of course this is a poor assumption. 

If we have only one library that has DB knowledge and exposes various operations, and multiple daemons that import/use it, this could give us good separation of concerns. I see your point. It just means a bit more work up front, but if we spec it right it might allow for more people to work in parallel on smaller chunks of functionality.

**Comment by lrvick** [comment]
I think we are potentially converging on creating:
1. A database abstraction library. Perhaps overhaul provisor but provide the same general operations
2. A service that provides user CRUD operations over a socket we can run locally anywhere needed.
3. An HTTP API service that provides any general operations we want exposed to the world. User creation, stats, etc

Yes?

**Comment by KellerFuchs** [comment]
No.

My point was that we have several bits of functionallity that have strongly different needs regarding the DB (both in terms of the abstraction and in terms of their access level) and have different requirements regarding who can access it:
- user creation: `INSERT` only (or whatever is the equivalent in the backing store), accessible only by the user creation frontend (i.e. the SSH thing);
- hashbangctl: `SELECT` or `UPDATE`, properly authorizes rw access, accessible locally (over a UNIX socket);
- stats: access to some specific columns (or even to a DB view?), accessible globally (likely over an HTTP API).

While the user creation “API frontend” and hashbangctl can probably share a lib/DB abstraction (and that might be the right place to validate the data that is associated to users), stats definitely has different needs.

All the other things we mentionned can be built on top of the APIs provided by those 3.

**Comment by KellerFuchs** [comment]
@lrvick Is that something we agree on?

**Comment by daurnimator** [comment]
We don't need to create a new database here; or a new protocol. Just use postgres.....

**Comment by KellerFuchs** [comment]
@daurnimator Where did anyone suggest inventing a DB?

**Comment by daurnimator** [comment]
I feel like lrvick was starting to consider it.

Otherwise my comment was replying to:

> Can you explain why we would want Postfix/DNS to talk directly to the DB instead of being abstracted via socket/http interfaces? Imo the clients should not need to know our database type, schema or anything, 

==> Why invent a new protocol when one already exists?
And we _will_ have a schema. Better to make it a proper SQL one than some adhoc thing only in documentation.

If we want a http interface throw [postgrest](http://postgrest.com) on top.

**Comment by KellerFuchs** [comment]
@hashbang/administrators @hashbang/contributers 
I wrote [a design proposal](https://gist.github.com/KellerFuchs/a71ef61f7834d10473cb22abdbd1ba51). All comments are welcome, both on the requirements and the proposed design.

It's currently barely more than a brain-dump, as far as writing goes, but it contains the important info (and DB schemata!).

**Comment by RyanSquared** [comment]
@hashbang/administrators was this completed when I went on break?

**Comment by KellerFuchs** [comment]
@ChickenNuggers Not yet.
- [x] The [userdb design](https://github.com/hashbang/userdb) is nailed down.
- [x] The DB schema and configuration for various integrations (NSS and Postfix) are written and have unit tests.
- [ ] The API server and #!ctl rewrite are still pending.
- [ ] We need to document the migration procedure and write the script, but basically
  - Disable user creation and #!ctl modification.
  - Copy all user records to the new userdb.
  - Update the servers (one by one?) to use the new userdb in the backend.
  - Turn off the LDAP container, start the new API server.

The main blocker that @lrvick and I have been grinding at is indirect: we need to be able to reproduce our infra (or at least the shell servers) to test all of that before going live.
